services/user/Registry/service/src/lib.rs

```diff
@@ -0,0 +1,697 @@
+                check(
+                    Url::parse(uri).is_ok(),
+                    format!("Invalid redirect URI format: {}", uri).as_str(),
+                );
+            }
+
+            for owner in &self.owners {
+                check(
+                    AccountsSvc::call().exists(*owner),
+                    format!("Owner account does not exist: {}", owner).as_str(),
+                );
+            }
+
+            if self.status == AppStatus::Published as AppStatusU32 {
+                let publishing_required_fields = [
+                    ("account_id", &self.account_id.to_string()),
+                    ("name", &self.name),
+                    ("short_description", &self.short_description),
+                    ("long_description", &self.long_description),
+                    ("tos_subpage", &self.tos_subpage),
+                    ("privacy_policy_subpage", &self.privacy_policy_subpage),
+                    ("app_homepage_subpage", &self.app_homepage_subpage),
+                ];
+                for (field_name, field_value) in publishing_required_fields {
+                    check(
+                        field_value.len() > 0,
+                        format!("{} is required for published apps", field_name).as_str(),
+                    );
+                }
+
+                // TODO: check each of those subpages exist
+                // in the caller's namespace in sites.
+            }
+        }
+    }
+
+    #[table(name = "AppTagsTable", index = 3)]
+    #[derive(Debug, Clone, Fracpack, ToSchema, Serialize, Deserialize, SimpleObject)]
+    pub struct AppTag {
+        /// The unique identifier for the app
+        pub app_id: AccountNumber,
+
+        /// The unique identifier for the tag
+        pub tag_id: u32,
+    }
+
+    impl AppTag {
+        #[primary_key]
+        fn by_app_tag_ids(&self) -> (AccountNumber, u32) {
+            (self.app_id, self.tag_id)
+        }
+
+        #[secondary_key(1)]
+        fn by_tag_id_apps(&self) -> (u32, AccountNumber) {
+            (self.tag_id, self.app_id)
+        }
+    }
+
+    #[derive(SimpleObject, Pack, Unpack, Deserialize, Serialize, ToSchema)]
+    pub struct AppMetadataWithTags {
+        pub metadata: AppMetadata,
+        pub tags: Vec<TagRecord>,
+    }
+
+    #[derive(SimpleObject, Pack, Unpack, Deserialize, Serialize, ToSchema)]
+    pub struct RelatedTags {
+        pub tags: Vec<String>,
+    }
+
+    #[action]
+    fn init() {
+        let table = InitTable::new();
+        table.put(&InitRow {}).unwrap();
+
+        EventsSvc::call().setSchema(create_schema::<Wrapper>());
+        EventsSvc::call().addIndex(
+            DbId::HistoryEvent,
+            SERVICE,
+            MethodNumber::from("appMetadataChanged"),
+            0,
+        );
+        EventsSvc::call().addIndex(
+            DbId::HistoryEvent,
+            SERVICE,
+            MethodNumber::from("appStatusChanged"),
+            1,
```

Comment:
I think what you want:
* IS a historical view of all status changes for a particular app
* is NOT a historical view of all apps that historically had a particular status

Right? If so, it implies that you want col 0 to be the index for this `appStatusChanged` event.